eth-contracts/contracts/service/ServiceProviderFactory.sol

```diff
@@ -248,23 +236,31 @@ contract ServiceProviderFactory is RegistryContract {
         // Update endpoint mapping
         serviceProviderEndpointToId[keccak256(bytes(_endpoint))] = 0;
 
+        require(
+          keccak256(bytes(serviceProviderInfo[_serviceType][deregisteredID].endpoint)) == keccak256(bytes(_endpoint)),
+          "Invalid endpoint for service type");
+
         require (
             serviceProviderInfo[_serviceType][deregisteredID].owner == msg.sender,
-            "Invalid deregister operation");
+            "Only callable by endpoint owner");
 
         // Update info mapping
         delete serviceProviderInfo[_serviceType][deregisteredID];
         // Reset id, update array
         uint spTypeLength = serviceProviderAddressToId[msg.sender][_serviceType].length;
+        bool idFound = false;
         for (uint i = 0; i < spTypeLength; i ++) {
             if (serviceProviderAddressToId[msg.sender][_serviceType][i] == deregisteredID) {
                 // Overwrite element to be deleted with last element in array
                 serviceProviderAddressToId[msg.sender][_serviceType][i] = serviceProviderAddressToId[msg.sender][_serviceType][spTypeLength - 1];
                 // Reduce array size, exit loop
                 serviceProviderAddressToId[msg.sender][_serviceType].length--;
+                // Confirm this ID has been found for the service provider
+                idFound = true;
                 break;
             }
         }
+        require(idFound == true, "Failed to find service type and ID for owner");
```

Comment:
after the first require statement for _endpoint, aren't all the rest of these basically asserts? like they should be unreachable? in sense that if any one of these statements breaks then something has seriously gone wrong?

Reply:
will remove